Remove ICustomPacket and add PayloadChannel.

LexManos opened the pull request:
Removal of ICustomPacket in favor of just using a custom Payload type. This drastically simplifies the patches, as I no longer have to edit the vanilla DiscardedPayload or Custom packets.
Connection now tracks and exposes both Inbound and Outbound ProtocolInfo, just useful for utilities like PacketLogger.

The new PayloadChannel is a copy of SimpleChannel but instead of using Classes to identify the packets it forces them to be subclasses of CustomPacketPayload and uses CustomPacketPayload.Type to identify.
They are still sent over the wire using a varint discriminator, because redundant info.

I added a generic `channel(Function)` method to ChannelBuilder to allow modders to create their own channels however they want. So that we don't have to do these things in the future.

LexManos committed:
More Network cleanup. Patch simplification

LexManos committed:
Add PayloadChannel it's just SimpleChannel but uses CustomPacketPayload.Type as the identifier instead of the class.

LexManos committed:
Small cleanup

PaintNinja commented on src/main/java/net/minecraftforge/network/PayloadChannel.java:
```diff
@@ -0,0 +1,289 @@
+        private boolean built = false;
+        private int lastIndex = 0;
+        protected Int2ObjectMap<Message<?, ?>> byId = new Int2ObjectArrayMap<>();
+        protected Object2ObjectMap<CustomPacketPayload.Type<?>, Message<?, ?>> byType = new Object2ObjectArrayMap<>();
```
This Map is updated by the add(Message) method but never queried or used outside of that

LexManos replied:
Right, meant to copy them in the channel constructor. Didn't write a test of this code so didn't catch that.

PaintNinja commented on src/main/java/net/minecraftforge/common/util/PacketLogger.java:
```diff
@@ -33,20 +39,39 @@ public PacketLogger(Connection connection) {
+        } else if (packet instanceof ServerboundCustomQueryAnswerPacket custom) {
+            channel = packet.getClass().getName() + " id " + custom.transactionId();
+            hex = hex(protocol, packet);
+        }
```
This can be a pattern matching switch in Java 21, like this:
```java
switch (packet) {
    case ClientboundCustomPayloadPacket custom -> {
        channel = custom.payload().type().id().toString();
        hex = hex(protocol, packet);
    }
    case ServerboundCustomPayloadPacket custom -> {
        channel = custom.payload().type().id().toString();
        hex = hex(protocol, packet);
    }
    case ClientboundCustomQueryPacket custom -> {
        channel = custom.payload().id().toString() + " id " + custom.transactionId();
        hex = hex(protocol, packet);
    }
    case ServerboundCustomQueryAnswerPacket custom -> {
        channel = packet.getClass().getName() + " id " + custom.transactionId();
        hex = hex(protocol, packet);
    }
    default -> {}
}
```

PaintNinja commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -1110,20 +1115,54 @@ public static void onCreativeModeTabBuildContents(CreativeModeTab tab, ResourceK
-            return false;
-        }
+    @Nullable
+    public static <B extends FriendlyByteBuf> StreamCodec<B, ? extends CustomPacketPayload> getCustomPayloadCodec(ResourceLocation id, int max) {
```
Marked as nullable but I can't see where it would return null... possible mistake?

PaintNinja commented on src/main/java/net/minecraftforge/network/PayloadChannel.java:
```diff
@@ -0,0 +1,289 @@
+        }
+    }
+
+    private interface ProtocolFactory<BUF extends FriendlyByteBuf, BASE extends CustomPacketPayload> extends PayloadConnection<BASE>, PayloadBuildable {
```
Possible candidate for `sealed` keyword (`permits` can be omitted in this case)

LexManos replied:
It's private, what would the benefit of sealed without permits be?

PaintNinja commented on src/main/java/net/minecraftforge/network/PayloadChannel.java:
```diff
@@ -0,0 +1,289 @@
+        int index,
+        Type<MSG> type,
+        NetworkProtocol<BUF> protocol,
+        PacketFlow direction,
```
These two record components can be marked with `@Nullable` as they're sometimes fed nulls and have null handling

PaintNinja commented on src/main/java/net/minecraftforge/event/network/CustomPayloadEvent.java:
```diff
@@ -26,27 +26,30 @@
+    public CustomPayloadEvent(ResourceLocation channel, Object payload, Context source, int loginIndex) {
         this.channel = channel;
         this.payload = payload;
+        this.data = payload instanceof ForgePayload forge ? forge.data() : null;
```
Would it make sense to also `instanceof FriendlyByteBuf` and use directly if available instead of null?

LexManos replied:
No, the payload will always be an instance of the vanilla Custom*Payload types. The issue is that those types are disjointed and thus I can't pass it in as a specific type.

PaintNinja reviewed (changes requested):
The reduced amount of patches and PacketLogger utilities are nice, likewise for supporting `CustomPacketPayload.Type` and the generic channel method.

Added some general feedback/code review, but not experienced with the networking myself, so leaving the more architectural side of things to others like crayfish.

LexManos committed:
Address feedback

LexManos committed:
Rework channel registration to allow multiple vanilla payload types per single forge channel.
Implemnt the minecraft:register/unregister channels using the new PayloadChannel and multiple-payload per channel system.

LexManos committed:
Add any() to SimpleConnection

LexManos committed:
Make Channel.isRemotePresent check all associated channels, not just the main one.

MrCrayfish commented on src/main/java/net/minecraftforge/network/NetworkProtocol.java:
```diff
@@ -41,23 +40,26 @@ public ConnectionProtocol toVanilla() {
-    public <T extends Packet<?>, MSG> ICustomPacket<T> buildPacket(PacketFlow direction, Channel<MSG> channel, MSG packet) {
-        var name = channel.getName();
-        var encoder = (Consumer<FriendlyByteBuf>)(buf -> channel.encode(buf, packet));
+    public <T extends PacketListener, MSG> Packet<T> buildPacket(PacketFlow direction, ResourceLocation name, Consumer<FriendlyByteBuf> encoder) {
```
MSG is now unused here

MrCrayfish commented on src/main/java/net/minecraftforge/network/NetworkProtocol.java:
```diff
@@ -9,17 +9,16 @@
-import net.minecraft.network.protocol.login.custom.DiscardedQueryPayload;
+import net.minecraft.resources.ResourceLocation;
 
 public class NetworkProtocol<B extends FriendlyByteBuf> {
```
`B` is now unused, however I understand that removing this would break everything. Could this be incorporated into the consumer of `buildPacket`?

LexManos committed:
Some minor generic cleanups.
